plugins/multiplexer/fetcher.h

```diff
@@ -219,14 +220,17 @@ template <class T> struct HttpTransaction {
             available = self->chunkDecoder_->decode(self->in_->reader);
             if (available == 0) {
               self->t_.data(self->in_->reader, available);
```

Comment:
I'm not familiar with this plugin, so I'd clarify a bit. What this `t_.data()` does?

Is it ok, calling `TSIOBufferReaderAvail` *after* the `t_.data()`? The order of the call should be like this?
```
              available = TSIOBufferReaderAvail(self->in_->reader);
              self->t_.data(self->in_->reader, available);
              TSIOBufferReaderConsume(self->in_->reader, available);
```

Reply:
It looks like if the `available` is larger than actual `read_avail()`, the `length += l;` will be problem.
https://github.com/apache/trafficserver/blob/7bea1d655e92ee62a6ffd997a457c9338c3f60ef/plugins/multiplexer/dispatch.cc#L185

Reply:
That `length` is used for a statistic. And it is fine to increment it by `l` here because the `available` value passed in is correct at that point in the code.

The problem is that the `data()` call, after the increment mentioned, may read from the reader. Which is fine, but then `available` in the calling scope obviously must change in order for future calls to `data()` to be accurate. Thus the update to `available` after the call to `data()`.

Reply:
> available value passed in is correct at that point in the code

How can you verify that it is always correct at this point in the code? Does it make sense to do some verification of `l` against the buffer reader `r`?

Reply:
I suppose we can add a debug assertion. To me the code reads that the implied contract is that `l` is less than or equal to the number of available bytes in the reader.

I'll add a debug assertion.

Thanks for the review.

Reply:
Actually, digging into this further, I notice that the issue with multiplexer is not `data()` at all but pulling bytes from the reader for parsing the headers. I've updated (the much simpler) patch.

Thanks again for the review @serrislew !